### ISSUE TITLE
seeking feedback: debugging-aid

### ISSUE THREAD

**Issue by naugtur** [nodejs/diagnostics#324]
I'm experimenting with possible ways of debugging and would like to discuss their usefulness and validity.

https://github.com/naugtur/debugging-aid

The tools use not-so-well-known features of Node.js to help figure out common problems.

- promises: global promise issues handlers are fairly obvious to everyone here, it just logs them
- blocked: blocked-at has already been discussed here somewhere I think. It uses async hooks to get a stack trace pointing to the blocking function
- network: network tab in debugger is something I've always wanted, this is the closest thing
- moduse: adds a getter/setter on Module._cache for a given id and uses that to track where it was required. I wonder if `require` checking the cache for built-in modules as well is something I can rely on. Got the idea for this when wondering how to detect if a malicious library loads `net` to call home.

**Comment by mhdawson** [comment]
@naugtur would it make sense for you to give a presentation/walk through these tools in a Diagnostics WG meeting?

**Comment by naugtur** [comment]
As long as it's the one tomorrow, I can do it.

**Comment by naugtur** [comment]
I presented this at the recent Diagnostics WG meeting and had some questions. People suggested I ask @mcollina these two:

**TL;DR bolded**

**1.** One of my experiments (moduse.js) hooks into Module._cache with a getter and setter and prints stack traces leading to each place in the codebase where a certain module is required. I built it for the specific purpose of finding out which packages are requiring the 'net' or 'http' internal modules.

I was surprised it actually works that way - **when you require('net') cache is checked even though it's a built-in module** and the cache is never populated.

**Is that desired behavior** for some obvious reason I'm missing? I'm considering adding a PR with a test that would assert this and asking if the assertion should require or forbid this behavior. The tests for module loading I found didn't cover this case.

**2.** Another experiment is using perf tracing for output. This question is rather simple.
**Is there a way to get node_trace.1.log to get written without performing a graceful shutdown? Can I trigger writing it?**

I've rarely seen http servers written in Node.js with logic to gracefully shutdown by closing the sockets and removing all intervals.
There might be a threshold of content at which the file is written while the process is still running, but what if I'm running it locally and there's very little output?

**I'm offering to turn the answers I get into documentation ;)**

**Comment by mcollina** [comment]
> 1. One of my experiments (moduse.js) hooks into Module._cache with a getter and setter and prints stack traces leading to each place in the codebase where a certain module is required. I built it for the specific purpose of finding out which packages are requiring the 'net' or 'http' internal modules.
>
> I was surprised it actually works that way - when you require('net') cache is checked even though it's a built-in module and the cache is never populated.

Some modules (https://www.npmjs.com/package/proxyquire) use that mechanism to intercept module loading to place mocks.

> Is there a way to get node_trace.1.log to get written without performing a graceful shutdown? Can I trigger writing it?

I do not know the answer to this.

**Comment by vmarchaud** [comment]
> Is there a way to get node_trace.1.log to get written without performing a graceful shutdown? Can I trigger writing it?

You can use the `inspector` module, i've a working implementation [here](https://github.com/vmarchaud/openprofiling-node/blob/master/packages/openprofiling-inspector-trace-events/src/inspector-trace-events-profiler.ts).
If you are interested in a minimum sample, look at this unit test: https://github.com/nodejs/node/blob/master/test/parallel/test-trace-events-dynamic-enable.js

**Comment by QuantumInformation** [comment]
Any video demos of this? Just want the network tab.

**Comment by naugtur** [comment]
@vmarchaud thanks, will check it out.
@QuantumInformation I demoed this on a Diag WG meeting before posting this thread and the recording should be in the archive, but you'll save a lot of time by reading the readme on the repo and just trying it out. It prints to standard output. You'll see.
